### PR TITLE
add vald shortlist parsing as well as handling for wavelength in vacuum/air and nm/aa

### PR DESCRIPTION
**Pull request by jvshields**
### :pencil: Description

**Type:** : :rocket: `feature` | 

This PR modifies the vald parser to add some extra functionality for more varied vald outputs. 

There are some missed lines in codecov to account for if the units of the wavelengths are in nm instead of angstroms, but I can't cover them unless I upload more data files and I've been trying to be sparing with that. I can upload them and cover them if that'd be preferred. 

### :vertical_traffic_light: Testing

- [x] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)

### :ballot_box_with_check: Checklist

- [ ] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

### CODE REVIEW

File: carsus/io/vald/vald.py [hunk 1]
```diff
@@ -193,38 +249,39 @@ def extract_linelist(self, vald):
                 vald linelist containing only the following columns:
                 atomic_number or chemical, ion_charge, wavelength, log_gf, rad, stark, waals
```
Comment (andrewfullard) [inline carsus/io/vald/vald.py:250]: e_low missing from this list

Comment (andrewfullard) [reply carsus/io/vald/vald.py:250]: Do we have a more complete description of what these values mean, too? This applies to other similar docstrings. They do not need to be in the docstrings, but *somewhere* in Carsus docs.

Comment (jvshields) [reply carsus/io/vald/vald.py:250]: Good catch. Updating the docstrings now. 

Where's the best place to describe what these values mean? I don't immediately know where makes the most sense.

Comment (andrewfullard) [reply carsus/io/vald/vald.py:250]: I would suggest adding a notebook to the datasets section of the docs with the info and possibly examples.

**Review by andrewfullard — APPROVED** [review]
Love the notebook

File: carsus/io/tests/test_vald.py [hunk 2]
```diff
@@ -115,3 +127,28 @@ def test_vald_linelist(vald_linelist):
...
+        (17, 5001.397869850396, -6.421, 7.1801, 1, 0),
+    ],
+)
+def test_vald_short_stellar_linelist(
+    vald_linelist_short_stellar, index, wavelength, log_gf, e_low, v_mic, ion_charge
+):
+    assert len(vald_linelist_short_stellar) == 95
+    row = vald_linelist_short_stellar.iloc[index]
+    assert_almost_equal(row["wavelength"], wavelength)
+    print(row)
```
Comment (Knights-Templars) [inline carsus/io/tests/test_vald.py:145]: Do you need this print statement here?

File: carsus/conftest.py [hunk 3]
```diff
@@ -137,6 +137,11 @@ def vald_fname(data_dir):
     return os.path.join(data_dir, "valdtest.dat")
 
 
+@pytest.fixture(scope="session")
+def vald_short_stellar_fname(data_dir):
+    return os.path.join(data_dir, "vald_shortlist_test.dat")
```
Comment (Knights-Templars) [inline carsus/conftest.py:142]: `short format`

File: carsus/io/vald/vald.py [hunk 4]
```diff
@@ -25,19 +26,23 @@ class VALDReader(object):
...
         Return pandas DataFrame representation of vald
+    linelist:
+        Return pandas DataFrame representation of linelist properties necessary to compute line opacities
 
     """
 
     vald_columns = [
         "elm_ion",
-        "wl_air",
+        "wave_unprepared",
```
Comment (Knights-Templars) [inline carsus/io/vald/vald.py:45]: Make comment

File: carsus/io/vald/vald.py [hunk 5]
```diff
@@ -129,12 +156,31 @@ def read_vald_raw(self, fname=None):
...
+            if "Vmicro" in line and self._stellar_linelist == False:
+                logger.info("Found Vmic column - This is a stellar vald linelist")
+                self._vald_columns.insert(3, "v_mic")
+
+                self._stellar_linelist = True
+
         vald = pd.read_csv(
-            StringIO("\n".join(data_match.findall(buffer.read().decode()))),
-            names=self.vald_columns,
+            StringIO("\n".join(data_match.findall(content))),
```
Comment (Knights-Templars) [inline carsus/io/vald/vald.py:176]: `data_match` as a constant and give it a better name.

File: carsus/io/vald/vald.py [hunk 6]
```diff
@@ -191,40 +249,42 @@ def extract_linelist(self, vald):
...
+                "chemical",
+                "ion_charge",
+                "wavelength",
+                "log_gf",
+                "e_low",
+                "rad",
+                "stark",
+                "waals",
+            ]
+            if self._stellar_linelist:
```
Comment (Knights-Templars) [inline carsus/io/vald/vald.py:266]: Please add a comment!

Comment (jvshields) [reply carsus/io/vald/vald.py:266]: It's actually already in the docstring above explaining. I think adding it here is redundant?

**Review by Knights-Templars — APPROVED** [review]

File: carsus/io/vald/vald.py [hunk 7]
```diff
@@ -158,7 +203,19 @@ def parse_vald(self, vald_raw=None, strip_molecules=True):
         vald["elm_ion"] = vald["elm_ion"].str.replace("'", "")
         vald[["chemical", "ion_charge"]] = vald["elm_ion"].str.split(" ", expand=True)
         vald["ion_charge"] = vald["ion_charge"].astype(int) - 1
-        vald["wavelength"] = convert_wavelength_air2vacuum(vald["wl_air"])
+
+        wave = vald.columns[1]
+        if "nm" in wave:
```
Comment (wkerzendorf) [inline carsus/io/vald/vald.py:208]: that should be a seperate function. that gives back air/vacuum and unit

File: carsus/tests/data/vald_shortlist_test.dat [hunk 8]
```diff
@@ -0,0 +1,100 @@
+ WARNING: Output was truncated to 100000 lines
+ 5000.00000, 5020.00000, 100000, 100001, 1.0 Wavelength region, lines selected, lines processed, Vmicro
+                                                 Damping parameters   Lande  Central
+Spec Ion       WL_air(A) Excit(eV) Vmic log gf*  Rad.   Stark  Waals  factor  depth  Reference
```
Comment (wkerzendorf) [inline carsus/tests/data/vald_shortlist_test.dat:4]: a function that takes this `WL_air(A)` and transforms it to enum of air and vacuum (or bool) and then to unit

File: carsus/conftest.py [hunk 9]
```diff
@@ -103,7 +103,7 @@ def memory_session():
 
 @pytest.fixture(scope="session")
 def data_dir():
-    return os.path.join(os.path.dirname(__file__), "tests", "data")
+    return Path(__file__).parent / "tests" / "data"
```
Comment (wkerzendorf) [inline carsus/conftest.py:106]: I would not make this a fixture but a single constant `DATA_DIR_PATH`

File: carsus/io/vald/vald.py [hunk 10]
```diff
@@ -126,17 +152,58 @@ def read_vald_raw(self, fname=None):
...
 
-        data_match = re.compile("'[a-zA-Z]+ \d+',[\s*-?\d+[\.\d+]+,]*")
+        DATA_RE_PATTERN = re.compile("'[a-zA-Z]+ \d+',[\s*-?\d+[\.\d+]+,]*")
 
         buffer, checksum = read_from_buffer(self.fname)
+        content = buffer.read().decode()
+
+        # Need to identify the wavelength column header and overwrite the wavelength to obtain units and air or vacuum
+        # Also need to identify if Vmic is in the columns for correct column construction
+        for line in content.split("\n")[:10]:
```
Comment (wkerzendorf) [inline carsus/io/vald/vald.py:162]: make an issue about pattern matching.

**Comment by jvshields** [comment]
Created issue #389 as a future project idea.

**Pull request merged**